lib/pure/collections/lists.nim

```diff
@@ -523,6 +591,72 @@ proc prepend*[T](L: var DoublyLinkedList[T], value: T) =
     assert a.contains(9)
   prepend(L, newDoublyLinkedNode(value))
 
+func copy*[T](a: DoublyLinkedList[T]): DoublyLinkedList[T] {.since: (1, 5, 1).} =
+  ## Creates a shallow copy of `a`.
+  runnableExamples:
+    type Foo = ref object
+      x: int
+    var f = Foo(x: 1)
+    let
+      a = [f].toDoublyLinkedList
+      b = a.copy
+    f.x = 42
+    assert a.head.value.x == 42
+    assert b.head.value.x == 42
+
+    let c = [1, 2, 3].toDoublyLinkedList
+    assert $c == $c.copy
+  result = initDoublyLinkedList[T]()
+  for x in a:
+    result.append(x)
+
+proc addMoved*[T](a, b: var DoublyLinkedList[T]) {.since: (1, 5, 1).} =
+  ## Moves `b` to the end of `a`. Efficiency: O(1).
+  ## Note that `b` becomes empty after the operation.
+  ## Self-adding results in an empty list.
+  ##
+  ## See also:
+  ## * `add proc <#add,T,T>`_
+  ##   for adding a copy of a list
+  runnableExamples:
+    import sequtils
+    var
+      a = [1, 2, 3].toDoublyLinkedList
+      b = [4, 5].toDoublyLinkedList
+    a.addMoved b
+    assert a.toSeq == [1, 2, 3, 4, 5]
+    assert b.toSeq == []
+    a.addMoved a
+    assert a.toSeq == []
+  if b.head != nil:
+    b.head.prev = a.tail
+  if a.tail != nil:
+    a.tail.next = b.head
+  a.tail = b.tail
+  if a.head == nil:
+    a.head = b.head
+  b.head = nil
```

Comment:
change to:
```nim
  if a.addr != b.addr:
    b.head = nil
    b.tail = nil
```

and add this test:
```nim
  import lists, sequtils
  import std/enumerate
  import std/sugar
  var a = [0,1].toSinglyLinkedList # actualy, `toList` in your test
  a.addMoved a
  let s = collect:
    for i, ai in enumerate(a):
      if i >= 6: break
      ai
  doAssert s == [0,1,0,1,0,1]
```

and edit doc comment to:
```nim
  ## Note that `b` becomes empty after the operation.
  ## Self-adding results in an empty list.
```
=>
```nim
  ## Note that `b` becomes empty after the operation unless it has same address as `a`.
```

that's IMO a saner behavior

* ditto with SinglyLinkedList

Reply:
If I change just this, it will create a circular list. I assume that you meant that self-addition should be a no-op, right?